### PR TITLE
[FIX] account: do not create invoices for the future

### PR DESCRIPTION
**Pull request by ap-wtioit**
Description of the issue/feature this PR addresses:
* currently the tests for account generate invoices for the future (if current day of the month is below 8 or 15), invoices should not be generated for the future (they are only allowed to be generated for the past)

Current behavior before PR:
* Test generates invoices for the future

Desired behavior after PR is merged:
* Test generates invoices for the recent past / same day

Info @wt-io-it

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Yenthe666** [comment]
Does it really matter or break anything though? I don't see the real problem in it.

**Comment by gdgellatly** [comment]
Why should invoices only be allowed for the past?  Here is a real basic common example.  Trade terms that specify that goods dispatched on last day of month, must be invoiced on 1st day of following month (when they receive them, days after EOM terms).  But there are hundreds of other common reasons to forward date an invoice.

**Comment by mart-e** [comment]
Why is that an issue?
It should target master if it is only cleaning of demo data and anyway, if it does not raise any issue, I don't see the problem, it is nice to have invoices at different dates for the demonstration.

**Comment by gdgellatly** [comment]
@mart-e Its more than nice, it means that for at least half the month we are checking that forward dating invoices work.

**Comment by wtaferner** [comment]
Alright, haha, I somehow said that it will never be accepted in 10.0.

Basically it is owed to the fact that it is actually almost nowhere allowed to validate invoices on a specific date in the future nor in the past (as long you do not mess up the sequences and dates it is not an issue IMHO, but yeah, there are plenty of legislation's around the world, so maybe my perception is wrong) 
So, @gdgellatly you are right, that you can create invoices in the future in certain use cases but normally you would do this on the day of issuance and if not in an own sequence and you will not block your own sequence as normally it must be clean by number and date ascending afaik.

So, if we want to test the cases Graeme we should explicitly test them but not by creating draft invoices with a date as it will almost fail every protection on the ```invoice_validate``` which is legit.

So, long story short, if it is creating too much heat and it does not have an issue in core, master is the way to go for sure, but we somehow hoped that this nasty, graved in stone test data can be skipped as no real test is depending on it.

@qdp-odoo 
I would be interested if you agree that it is no issue to do outgoing invoices in the future or the past and basically suggest this via test data, even it is possible under several circumstances, no?

**Comment by gdgellatly** [comment]
@wtaferner I know some countries have regulations like that. But that is localisation specific.

Here is the NZ tax rules - no sequence even required http://www.ird.govt.nz/yoursituation-bus/running/recordkeeping/recordkeeping-taxinvoiceillustration.html

If you need an external auditor it is a good idea generally as an internal control and sequences are almost entirely used as common business practice, but definitely no requirement to keep in date order.

AFAICR same in Australia and UK.  I guess we are almost nowhere by your reckoning.

**Comment by wtaferner** [comment]
@gdgellatly 
Haha, well, obviously the tax authorities have too much money there that they allow to issue invoices whenever you want in order to have the "statistics" you would like to have at some point ;-)

Hmm, anyway, if it is a localization thing the test should not be like this in core as it will break the ones who do care about invoices to be dated on the issuance date instead of the future or the past.

Thank you for the pointers and no, I do not do a study if my reckoning of **almost nowhere** would be correct ;-)

**Comment by gdgellatly** [comment]
@wtaferner Actually we have probably one of the best tax systems in the world as tax systems go.  Simple and easy to comply with, reasonably progressive, reasonably fair, well enforced, very little fraud.  

You have to remember that if I forward date an invoice then while I might defer some revenue, my customer must defer some expense, so the tax take is even. Only 1 tax rate for companies here, and it is an integrated tax system anyway, so companies are only really paying tax on behalf of shareholders until dividends are paid.  In the meantime, the forward dating has delayed my cashflows by 54 days usually.  We don't exactly like forward dating invoices, just what needs to be done sometimes.

**Comment by wtaferner** [comment]
@mart-e 
Do you agree that we should skip this date back and forth validation test in core and it should be tested in localization if there is a need to? 

As this is a YAML test it will kill the tests during the import phase of the module, so no tests will be executed after this one.
